src/Handlers.jl

```diff
@@ -17,25 +17,69 @@ function handle end
 """
 Abstract type representing an object that knows how to "handle" a server request.
 
-Types of handlers include `HandlerFunction` (a julia function of the form `f(request)` and
-`Router` (which pattern matches request url paths to other specific `Handler` types).
+Types of handlers include:
+  * `HTTP.RequestHandlerFunction`: a julia function of the form `f(request::HTTP.Request)`
+  * `HTTP.Router`: pattern matches request url paths to other specific `Handler` types
+  * `HTTP.StreamHandlerFunction`: a julia function of the form `f(stream::HTTP.Stream)`
 """
 abstract type Handler end
+abstract type RequestHandler <: Handler end
+abstract type StreamHandler <: Handler end
+
+Handler(h::Handler) = h
+function Handler(f::Base.Callable)
+    req = applicable(f, Request())
+    strm = applicable(f, Stream(Request(), IOBuffer()))
+    if strm && !req
+        h = StreamHandlerFunction(f)
+    elseif req
+        h = RequestHandlerFunction(f)
+    else
+        throw(ArgumentError("$f function doesn't take an Request or HTTP.Stream argument"))
+    end
+    return h
+end
 
 """
-HandlerFunction(f::Function)
+RequestHandlerFunction(f::Function)
 
 A Function-wrapper type that is a subtype of `Handler`. Takes a single Function as an argument.
 The provided argument should be of the form `f(request) => Response`, i.e. it accepts a `Request` returns a `Response`.
 """
-struct HandlerFunction{F <: Function} <: Handler
+struct RequestHandlerFunction{F <: Function} <: RequestHandler
     func::F # func(req)
 end
-
-handle(h::HandlerFunction, req) = h.func(req)
+RequestHandlerFunction(f::RequestHandlerFunction) = f
 
 "A default 404 Handler"
-const FourOhFour = HandlerFunction(req -> HTTP.Response(404))
+const FourOhFour = RequestHandlerFunction(req -> Response(404))
+
+@inline function handle(h::RequestHandler, stream::Stream)
+    request::Request = stream.message
+    request.body = read(stream)
+    request.response::Response = handle(h, request)
+    request.response.request = request
+    startwrite(stream)
+    write(stream, request.response.body)
+    return
+end
+
+handle(h::RequestHandlerFunction, req::Request) = h.func(req)
+
+struct StreamHandlerFunction{F <: Function} <: StreamHandler
```

Comment:
Constrainting callables to be subtypes of `Function`,
is generally not beneficial,
It breaks Constructors and custom callable types.